Lib/importlib/_bootstrap_external.py

```diff
@@ -738,19 +739,32 @@ def _validate_timestamp_pyc(data, source_mtime, source_size, name,
 
     *name* is the name of the module being imported. It is used for logging.
 
+    *bytecode_path* is the path of the pyc file.
+
     *exc_details* is a dictionary passed to ImportError if it raised for
     improved debugging.
 
     An ImportError is raised if the bytecode is stale.
 
     """
-    if _unpack_uint32(data[8:12]) != (source_mtime & 0xFFFFFFFF):
+    timestamp = _unpack_uint32(data[8:12])
+    if timestamp != (int(source_mtime) & 0xFFFFFFFF):
         message = f'bytecode is stale for {name!r}'
         _bootstrap._verbose_message('{}', message)
         raise ImportError(message, **exc_details)
     if (source_size is not None and
         _unpack_uint32(data[12:16]) != (source_size & 0xFFFFFFFF)):
         raise ImportError(f'bytecode is stale for {name!r}', **exc_details)
+    if time.time() - source_mtime < 2:
+        try:
+            bytecode_mtime = self.path_stats(bytecode_path)['mtime']
+        except OSError:
+            pass
+        else:
+            if bytecode_mtime < source_mtime:
```

Comment:
It is still possible to get a stale bytecode if write the new source very quickly after creating the pyc file. Using `<=` instead of `<` here would be more reliable, but this breaks several existing tests which create bad pyc files very quickly after creating py files.

Using nanosecond precision may make all this more reliable without breaking existing tests, but this will require changing the public interface -- adding the `mtime_ns` key in `path_stats()`.

Reply:
Using  nanoseconds precision does not help.